File: vbnet/App_Code/Core/WebExHtmlExtensions.vb

Imports System.Collections.Generic
Imports System.Dynamic
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Web
Imports System.Web.Mvc
Imports System.Web.Mvc.Html

''' <summary>
''' Summary description for WebExHtmlExtensions
''' </summary>
Public Module WebExHtmlExtensions
    Public Const DefaultViewType As String = "webex:defaulviewtype"
    Public Const webexViewExtension As String = "webex:viewext"
    Public Function GetViewExtension(app As HttpApplicationStateBase) As String
        Dim ext = TryCast(app(webexViewExtension), String)
        If String.IsNullOrEmpty(ext) Then
            ext = "cshtml"
        End If

        Return ext
    End Function
    <System.Runtime.CompilerServices.Extension> _
    Public Function RenderModule(helper As HtmlHelper, [module] As IModule, Optional view As String = "", Optional moduleModel As Object = Nothing) As MvcHtmlString
        If [module] Is Nothing Then
            Dim viewName = String.Format("~/Views/Modules/{0}.{1}", view, GetViewExtension(helper.ViewContext.HttpContext.Application))
            Return helper.[Partial](viewName, New WebExModuleNotFoundModel(Nothing, viewName, moduleModel))
        End If

        Dim v = view
        If String.IsNullOrEmpty(v) Then
            v = [module].GetViewOfType(DefaultViewType)
        End If

        Dim modelModule = TryCast([module], IModuleWithModel)
        If modelModule Is Nothing Then
            Return New MvcHtmlString(v)
        Else
            If String.IsNullOrEmpty(v) OrElse v = DefaultViewType Then
                v = "index"
            End If

            Return helper.RenderModule(GetModuleName([module]), DirectCast(modelModule.Model, Object), v)
        End If
    End Function

    Private Function GetModuleName([module] As IModule) As String
        Dim t = [module].[GetType]()
        Return t.Name.Replace("Module", String.Empty)
    End Function
    <System.Runtime.CompilerServices.Extension> _
    Public Function RenderModule(helper As HtmlHelper, [module] As String, Optional model As Object = Nothing, Optional view As String = "index") As MvcHtmlString
        Dim viewName = String.Format("~/Views/Modules/{0}/{1}.{2}", [module], view, GetViewExtension(helper.ViewContext.HttpContext.Application))
        If helper.PartialViewExists(viewName) Then
            Return helper.[Partial](viewName, model)
        End If

        viewName = String.Format("~/Views/Modules/{0}.{1}", view, GetViewExtension(helper.ViewContext.HttpContext.Application))
        If helper.PartialViewExists(viewName) Then
            Return helper.[Partial](viewName, New WebExModuleNotFoundModel([module], viewName, model))
        Else
            If helper.PartialViewExists(String.Format("~/Views/Modules/index.{0}", GetViewExtension(helper.ViewContext.HttpContext.Application))) Then
                Return helper.[Partial](String.Format("~/Views/Modules/index.{0}", GetViewExtension(helper.ViewContext.HttpContext.Application)), New WebExModuleNotFoundModel([module], viewName, model))
            End If
        End If

        Return Nothing
    End Function
    <System.Runtime.CompilerServices.Extension> _
    Public Function RenderModule(helper As HtmlHelper, model As WebExModel, [module] As String, Optional view As String = "index", Optional moduleModel As Object = Nothing) As MvcHtmlString
        Dim res As Object
        If model.TryGetProperty([module], res) Then
            Dim m As IModule = TryCast(res, IModule)
            If m IsNot Nothing Then
                Return helper.RenderModule(m, view, moduleModel)
            Else
                Return helper.RenderModule([module], res, view)
            End If
        End If
        Return Nothing
    End Function
    <System.Runtime.CompilerServices.Extension> _
    Public Function RenderPartialIfExists(html As HtmlHelper, partialViewName As String) As MvcHtmlString
        If html.PartialViewExists(partialViewName) Then
            Return html.[Partial](partialViewName)
        End If

        Return Nothing
    End Function
    <System.Runtime.CompilerServices.Extension> _
    Public Function PartialViewExists(html As HtmlHelper, partialViewName As String) As Boolean
        Return ViewEngines.Engines.FindPartialView(html.ViewContext.Controller.ControllerContext, partialViewName).View IsNot Nothing
    End Function
    <System.Runtime.CompilerServices.Extension> _
    Public Function ViewExists(html As HtmlHelper, viewName As String, Optional master As String = Nothing) As Boolean
        Return ViewEngines.Engines.FindView(html.ViewContext.Controller.ControllerContext, viewName, master).View IsNot Nothing
    End Function
    <System.Runtime.CompilerServices.Extension> _
    Public Iterator Function RenderModulesViewOfType(helper As HtmlHelper, model As WebExModel, viewType As String, Optional getOrderWeight As Func(Of IModule, Integer) = Nothing) As IEnumerable(Of MvcHtmlString)
        For Each [module] In model.GetMultiViewModules().OrderBy(Function(it) If(getOrderWeight Is Nothing, 0, getOrderWeight(it)))
            Dim view = [module].GetViewOfType(viewType)
            If Not String.IsNullOrEmpty(view) Then
                Yield helper.RenderModule([module], view:=view)
            End If
        Next
    End Function
End Module

File: csharp/App_Code/Core/WebExHtmlExtensions.cs
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.Linq;
using System.Linq.Expressions;
using System.Web;
using System.Web.Mvc;
using System.Web.Mvc.Html;

/// <summary>
/// Summary description for WebExHtmlExtensions
/// </summary>
public static class WebExHtmlExtensions
{
    public const string DefaultViewType = "webex:defaulviewtype";
    public const string webexViewExtension = "webex:viewext";
    public static string GetViewExtension(HttpApplicationStateBase app)
    {
        var ext = app[webexViewExtension] as string;
        if (string.IsNullOrEmpty(ext))
            ext = "cshtml";

        return ext;
    }
    public static MvcHtmlString RenderModule(this HtmlHelper helper, IModule module, string view = "", object moduleModel = null)
    {        
        if (module == null)
        {
            var viewName = string.Format("~/Views/Modules/{0}.{1}", view, GetViewExtension(helper.ViewContext.HttpContext.Application));
            return helper.Partial(viewName, new WebExModuleNotFoundModel(null, viewName, moduleModel));
        }

        var v = view;
        if (string.IsNullOrEmpty(v))
            v = module.GetViewOfType(DefaultViewType);

        var modelModule = module as IModuleWithModel;
        if (modelModule == null)
        {
            return new MvcHtmlString(v);
        }
        else
        {
            if (string.IsNullOrEmpty(v) || v == DefaultViewType)
                v = "index";

            return helper.RenderModule(GetModuleName(module), (object)modelModule.Model, v);
        }
    }

    private static string GetModuleName(IModule module)
    {
        var t = module.GetType();
        return t.Name.Replace("Module", string.Empty);
    }
    public static MvcHtmlString RenderModule(this HtmlHelper helper, string module, object model = null, string view = "index")
    {
        var viewName = string.Format("~/Views/Modules/{0}/{1}.{2}", module, view, GetViewExtension(helper.ViewContext.HttpContext.Application));
        if (helper.PartialViewExists(viewName))
        {
            return helper.Partial(viewName, model);
        }

        viewName = string.Format("~/Views/Modules/{0}.{1}", view, GetViewExtension(helper.ViewContext.HttpContext.Application));
        if (helper.PartialViewExists(viewName))
        {
            return helper.Partial(viewName, new WebExModuleNotFoundModel(module, viewName, model));
        }
        else
        {
            if (helper.PartialViewExists(string.Format("~/Views/Modules/index.{0}", GetViewExtension(helper.ViewContext.HttpContext.Application))))
                return helper.Partial(string.Format("~/Views/Modules/index.{0}", GetViewExtension(helper.ViewContext.HttpContext.Application)), 
                    new WebExModuleNotFoundModel(module, viewName, model));
        }

        return null;
    }
    public static MvcHtmlString RenderModule(this HtmlHelper helper, WebExModel model, string module, string view = "index", object moduleModel = null)
    {
        object res;
        if (model.TryGetProperty(module, out res))
        {
            IModule m = res as IModule;
            if (m != null)
                return helper.RenderModule(m, view, moduleModel);
            else
                return helper.RenderModule(module, res, view);
        }
        return null;
    }
    public static MvcHtmlString RenderPartialIfExists(this HtmlHelper html, string partialViewName)
    {
        if (html.PartialViewExists(partialViewName))
            return html.Partial(partialViewName);

        return null;
    }
    public static bool PartialViewExists(this HtmlHelper html, string partialViewName)
    {
        return ViewEngines.Engines.FindPartialView(html.ViewContext.Controller.ControllerContext, partialViewName).View != null;
    }
    public static bool ViewExists(this HtmlHelper html, string viewName, string master = null)
    {
        return ViewEngines.Engines.FindView(html.ViewContext.Controller.ControllerContext, viewName, master).View != null;
    }
    public static IEnumerable<MvcHtmlString> RenderModulesViewOfType(this HtmlHelper helper, WebExModel model, string viewType,
        Func<IModule, int> getOrderWeight = null)
    {
        foreach (var module in model.GetMultiViewModules().OrderBy(it=>getOrderWeight == null?0:getOrderWeight(it)))
        {
            var view = module.GetViewOfType(viewType);
            if (!string.IsNullOrEmpty(view))
                yield return helper.RenderModule(module, view: view);
        }
    }
}